### ISSUE TITLE
openblas not built with full processor support

### ISSUE THREAD

**Issue by kalefranz** [ContinuumIO/anaconda-issues#1661]
_From @susilehtola on May 4, 2017 4:33_

The openblas library appears to have been built only for the Nehalem processor. However, openblas has runtime cpu detection, and so it should be built with support for all processors.

Also, it's not clear which version of the library you are shipping - is it the sequential library, a pthreads library, or an OpenMP library? It would be good to have all three.

You can have a look at how it's supposed to be built at the Fedora spec file
 https://src.fedoraproject.org/cgit/rpms/openblas.git/tree/openblas.spec

For the dynamic architecture support, the flags you need are e.g.
TARGET=CORE2 DYNAMIC_ARCH=1

_Copied from original issue: conda/conda#5215_

**Comment by kalefranz** [comment]
_From @mingwandroid on May 4, 2017 9:33_


> The openblas library appears to have been built only for the Nehalem processor

Thanks we should fix that.

> .. is it the sequential library, a pthreads library, or an OpenMP library?

We have a build matrix dimension for the blas implementation (either MKL or OpenBLAS). Unfortunately that is done by the old 'features' feature. This only allows for two options (feature on or feature off) so we cannot, as things stand, support 3 variants of OpenBLAS (though conda-build 3 - which is not quite ready yet - does have support for non-binary feature).

I am interested in why you want sequential, pthreads and OpenMP though. Can you detail your use-case? I get that sequential can be faster on single core processors, but hardly anyone uses those now, and to my mind, OpenMP vs pthreads is an implementation detail that should be hidden from the user. For packaging, we should pick whichever is the fastest on each platform.

> You can have a look at how it's supposed to be built at the Fedora spec file

That is just one project's version of how it is supposed to be built, they have very different constraints and requirements than those of the Anaconda Distribution (I have not seen a Fedora for macOS or Windows for example). Here's another project's version of the same: https://aur.archlinux.org/cgit/aur.git/plain/PKGBUILD?h=openblas

**Comment by susilehtola** [comment]
Well, if I only want to a sequential binary, then I want to use sequential libraries, too, to avoid any overhead from the parallellization.

I believe Pthreads is a bit faster then OpenMP parallellization... however, OpenMP is by far the more often used case, since it is so much easier to write than pthreads.

So if you're restricted to one flavor, I'd pick OpenMP, because
* it allows paralallelization
* it does not screw things up when called from OpenMP parallel regions in the user side code
* its thread number can be adjusted easily

**

For the Fedora stuff, I was mainly referring to the variants we ship on different platforms and the compile flags for the different variants. Note we also ship a 64-integer interface version for handling matrices larger than 4GB.

**Comment by mingwandroid** [comment]
OK thanks for the info. I guess the 64-bit integer version is not ABI compatible with the 32-bit integer version and a different define is needed when linking to it, is that correct?

**Comment by susilehtola** [comment]
Yeah it's not ABI compatible, because it uses INTEGER(8) instead of the conventional INTEGER(4) in the Fortran calls; INTEGER(4) is the reason why the old interface is limited to 4GB matrices.

For this reason Fedora ships two different libraries for every flavor - one with 32-bit integers and one with 64-bit integers.
